astropy_healpix/core.py

```diff
@@ -77,10 +84,92 @@ def level_to_nside(level):
     nside : int
         The number of pixels on the side of one of the 12 'top-level' HEALPix tiles.
     """
+    level = np.asarray(level, dtype=np.int64)
+
     _validate_level(level)
     return 2 ** level
 
 
+def nside_to_level(nside):
+    """
+    Find the HEALPix level for a given nside. (this is given by log2(nside))
+
+    This function is the inverse of `level_to_nside`.
+
+    Parameters
+    ----------
+    nside : int
+        The number of pixels on the side of one of the 12 'top-level' HEALPix tiles.
+        Must be a power of two.
+
+    Returns
+    -------
+    level : int
+        The level of the HEALPix cells
+    """
+    nside = np.asarray(nside, dtype=np.int64)
+
+    _validate_nside(nside)
+    return np.log2(nside)
+
+
+def uniq_to_levelipix(uniq):
+    """
+	Convert a HEALPix cell uniq number to its (level, ipix) equivalent
+
+    A uniq number is a 64 bit integer equals to : ipix + 4*(4**level). Please read
+    this `paper <http://ivoa.net/documents/MOC/20140602/REC-MOC-1.0-20140602.pdf>`_
+    for more details about uniq numbers.
+
+    Parameters
+    ----------
+    uniq : int
+        The uniq number of a HEALPix cell.
+
+    Returns
+    -------
+    level, ipix: int, int
+        The level and index of the HEALPix cell computed from ``uniq``.
+    """
+    uniq = np.asarray(uniq, dtype=np.int64)
+
+    level = ((np.log2(uniq//4)) // 2)
```

Comment:
Remove the outermost `()`, it's superfluous.

Reply:
At least on x86, we can do better in assembly using the bsr instruction. See [my implementation here](https://git.ligo.org/lscsoft/ligo.skymap/blob/master/src/bayestar_moc.c#L36-53).

Reply:
@lpsinger - Is the bottleneck the `np.log2` or the whole expression?

Are you really limited by the speed of this expression in your analyses?
Or can we keep `astropy.healpix` dead simple and avoid code that might not run everywhere?

Reply:
I haven't benchmarked it in a while. Let me check. My code also includes a fallback for non-x86 platforms that uses integer arithmetic, though I don't know if that is faster than `np.log2`.

Reply:
I looked around a bit, there is no integer log 2 in Numpy.

Probably `np.log2` isn't fast for integers in the first place, and certainly the expression `(np.log2(uniq//4)) // 2` will not be fast because of temp Numpy array copies. So the first thing to do if speed matters here would be to move this to C with a simple implementation of `ilog2`(not using asm).

@lpsinger and @tboch - Is this speed really needed here? Or would it be OK to keep the `level = (np.log2(uniq//4)) // 2` because it has acceptable performance?

Reply:
I am attaching a short benchmark program (note: rename to `test.c` to compile). It's interesting, log2 is actually an order of magnitude faster than the bit shift method, but the x86 bsrq method is an order of magnitude faster than it.

## gcc6 from MacPorts

```
$ gcc -O3 -std=gnu11 test.c && ./a.out
bitshift: 3.003659e-08 s
log2:     7.648268e-09 s
bsrq:     1.006322e-09 s
```

## clang-6.0 from MacPorts

```
$ clang -O3 -std=gnu11 test.c && ./a.out
bitshift: 3.008260e-08 s
log2:     7.997110e-09 s
bsrq:     2.159280e-09 s
```

[test.c.txt](https://github.com/astropy/astropy-healpix/files/2359000/test.c.txt)

Reply:
Explicit bit-twiddling is most likely not the fastest solution here. Assembler instructions or compiler intrinsics would be good, but unfortunately we are in Python and numpy doesn't provide something like Python's `bit_length()`.

Unfortunately `np.log2()` is problematic too, since it can be inaccurate: see https://github.com/numpy/numpy/issues/4787 . This particular problem may have been fixed, but adding 0.5 before taking the logarithm may add some extra safety.